services-core/src/solution_submission/retry.rs

```diff
@@ -399,26 +414,61 @@ mod tests {
             .times(1)
             .return_once(|_, _, _, _, _| {
                 async move {
-                    receiver.await.unwrap();
+                    submit_receiver.await.unwrap();
                     Ok(())
                 }
                 .boxed()
             });
-        sleep.expect_sleep().times(1).returning(|_| immediate!(()));
+        sleep.expect_sleep().return_once(move |_| {
+            cancel_sender.send(()).unwrap();
+            future::pending().boxed()
+        });
+
+        let args = Args {
+            batch_index: 1,
+            solution: Solution::trivial(),
+            claimed_objective_value: 1.into(),
+            gas_price_cap: 10.0,
+            nonce: 0.into(),
+            target_confirm_time: Instant::now(),
+        };
+        let retry = RetryWithGasPriceIncrease::with_sleep_and_now(
+            Arc::new(contract),
+            Arc::new(gas_price),
+            sleep,
+            util::default_now(),
+        );
+        let result = retry.retry(args, cancel_future).wait();
+        assert!(matches!(result, RetryResult::Submitted(Ok(()))));
+    }
+
+    #[test]
+    fn cancellation_completes() {
+        let (cancel_sender, cancel_receiver) = futures::channel::oneshot::channel();
+        let mut contract = MockStableXContract::new();
+        let mut gas_price = MockGasPriceEstimating::new();
+        let mut sleep = MockAsyncSleeping::new();
+
+        let cancel_future = async move {
+            cancel_receiver.await.unwrap();
+        }
+        .boxed();
         gas_price
             .expect_estimate_with_limits()
             .times(1)
-            .returning(|_, _| Ok(2.0));
+            .returning(|_, _| Ok(1.0));
         contract
             .expect_submit_solution()
             .times(1)
-            .return_once(|_, _, _, _, _| {
-                sender.send(()).unwrap();
-                immediate!(Err(nonce_error()))
-            });
-        sleep
-            .expect_sleep()
-            .returning(|_| future::pending().boxed());
+            .return_once(|_, _, _, _, _| future::pending().boxed());
+        sleep.expect_sleep().return_once(move |_| {
```

Comment:
ditto: if possible I'd move the cancellation signalling into the `expect_submit_solution` mock.